SW counter measures against HW level fault injection

tamasban opened the pull request:
Adding SW counter measures against fault injection attacks in HW. These attacks could be clock or power glitching, EM pulse, laser beam, temperature change.  The goal of the attacker is is to execute unauthenticated code on the device.  The expectation of the attacker that due to these injected faults the CPU will change its normal behaviour and in this way create a benefit for the attacker to bypass secure boot and image authentication. The potential differences in the CPU behaviour could be one or more instruction skip, modifying register content, corrupting memory read, corrupting instruction decoding, resetting bits in program status register. 

The SW counter measure are not guaranteeing 100% protection against all HW attack but make harder to bypass image authentication. The SW counter measures are disabled by default they can be enabled with MCUBOOT_FIH_PROFILE_* macros. Currently there are defined 4 protection level: none, moderate, medium, max. These are enabling more and more SW counter measures. These are only added to the generic part of MCUboot files in `boot/bootutil/src`. Currently only the critical call path is hardened where a corruption of a single `if (ret)` statement can be accept an unauthenticated image. The crypto (mbed-crypto, tinycrypt) libraries are independent projects from MCUboot so they are not changed. They might contains similar feature or not, need to be checked individually.

Main hardening code are in :

- boot/bootutil/include/bootutil/fault_injection_hardening.h
- boot/bootutil/src/fault_injection_hardening.c

RcColes commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,339 @@
+#define FIH_ENABLE_GLOBAL_FAIL
+#define FIH_ENABLE_CFI
+
+#elif defined(MCUBOOT_FIH_PROFILE_MEDIUM)
```
IMO Using both `MODERATE` and `MEDIUM` could be very confusing. I think something like `OFF / LOW / MEDIUM / HIGH` would be better.

adrianlshaw replied:
Agreed

tamasban replied:
Fixed.

utzig commented:
This needs rebase. Also I would appreciate if you could submit me links to papers/&c that explain why those changes are made the way they are!

tamasban commented:
> This needs rebase. Also I would appreciate if you could submit me links to papers/&c that explain why those changes are made the way they are!

Here is two white paper in the topic and also the QEMU based testing (will be another PR) can prove the usefulness of these patterns:

- https://www.cl.cam.ac.uk/~rja14/Papers/whatyouc.pdf

- https://www.riscure.com/uploads/2018/11/201708_Riscure_Whitepaper_Side_Channel_Patterns.pdf

soby-mathew commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,339 @@
+    return rc;
+}
+#else
+__attribute__((always_inline)) inline
```
minor: The `always inline` attribute is still an indication that a function must be inlined. Macros are the better way to ensure inlining. Same comment below.

RcColes replied:
There's a deliberate tradeoff between usability and performance going on here; Macros are certainly be best way to ensure things are always inlined but they come with some tradeoffs. They tend to cause issues with debuggers / debug symbols and also things like code-completion / code formatters etc. All the compilers I tested did respect the `always_inline`, but I do agree that it could be an issue in theory.

tamasban replied:
Your are right. So far they were inlined. I would keep in this way for simplicity. If one want to use this feature then the compiled code must be verified by test and visual inspection. So functions not be inlined should be caught by these steps.

soby-mathew commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,339 @@
+ */
+#ifdef FIH_ENABLE_CFI
+#define FIH_CFI_PRECALL_BLOCK \
+    fih_int _fih_cfi_saved_value = fih_cfi_get_and_increment()
```
Just some blue sky thinking, not sure whether it can be done or not. If the increment can be done based on the function being invoked i.e some way to hash the name of the function into a unique number, then we can confirm if an increment is called for a function call, then the decrement in the same function need to be invoked for it to be validated. 

This would ensure that if the call is subverted to a different location and the RET from that different function will fail the check on return back to the caller.

RcColes replied:
It would be an interesting idea, but maybe too difficult in practice. The C preprocessor does have access to the function name via `__func__`, but I'm not sure if you could evaluate a hash function or similar inside the preprocessor. I'd avoid doing the hash inside actual code, as that'd cause a fairly large size increase due to the inlining.

In general, most functions don't have an FIH_RET and it's pretty hard to control the value of register corruption (and hence jump location). More targeted attacks like ion beams could probably target the lower bits deliberately, which would allow jumps to be slightly more precise.

adrianlshaw replied:
This would be ideal. A similar alternative might be easier: take the function address and derive a number from it. 

It's worth noting that fault injection techniques are unlikely to produce precise values for a return address. If precise return addresses are possible then it is time to consider more comprehensive ROP/JOP protection, which I believe is beyond the scope of the attacks/adversaries considered in this PR.

adrianlshaw replied:
I see Raef beat me to it :)

tamasban replied:
Might `sizeof(__func__)` would add some randomness.

soby-mathew reviewed (commented):
just some suggestion. Otherwise looks good to me.

adrianlshaw commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,339 @@
+ * counter is decremented, and after the return it is verified that the counter
+ * has the same value as before this process. This can be used to verify that
+ * the function has actually been called.
+ *
```
There should be some description here that describes how this only provides a specific subset of CFI. This is checking that a specific sequence of calls are made within the same block and that the adversary is someone performing glitching. It does not cover other types of CFI such as ROP protection or JOP protection.

tamasban replied:
Fixed

nvlsianpu commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,343 @@
+void fih_cfi_validate(fih_int saved);
+void fih_cfi_decrement(void);
+
+/* Label for interacting with FI testing tool. Can be parsed from the elf file
```
```suggestion
/* Label for interacting with FIH testing tool. Can be parsed from the elf file
```

tamasban replied:
ack

nvlsianpu commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,343 @@
+        return ret; \
+    } while (0)
+
+/* Macro wrappers for functions - Even when the functions have zero body this
```
move macro doc right above the macro (after `ifdef FIH_ENABLE_CFI`)

tamasban replied:
ack

utzig commented:
testing some message in GH slack integration, will remove...

utzig commented:
Please rebase.

tamasban commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,343 @@
+ *       be secure for all compilers, but execution is going to be correct and
+ *       including them will certainly help to harden the code.
+ *
+ * DOUBLE_VARS makes critical variables into a tuple (x, x ^ msk). Then the
```
ack

tamasban commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,343 @@
+ * Has support for different measures, which can either be enabled/disabled
+ * separately or by defining one of the MCUBOOT_FIH_PROFILEs.
+ *
+ * NOTE: These constructs against fault injection attacks are not guaranteed to
```
There is another PR, which is a test suite on top of QEMU:
https://github.com/JuulLabs-OSS/mcuboot/pull/789

This can be used with any compiler to validate the hardening code. It does not provide a 100% test coverage against FI on the entire mcuboot binary. It testes only that code sections which are between the FIH_CALL_START and FIH_CALL_END macros.

tamasban commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,343 @@
+{
+    fih_int_validate(x);
+    return x.val;
+}
```
ack

utzig commented:
@tamasban You don't really need to keep `ack`ing, do you? I get one email for every time you ack something. Why not just fix the problem and push?

tamasban commented:
> @tamasban You don't really need to keep `ack`ing, do you? I get one email for every time you ack something. Why not just fix the problem and push?

understood, sorry I get to used to it in Gerrit workflow.

utzig commented on boot/bootutil/src/bootutil_misc.c:
```diff
@@ -107,6 +108,38 @@ static const struct boot_swap_table boot_swap_tables[] = {
+ *
+ * @return      FIH_SUCCESS if memory regions are equal, otherwise FIH_FAILURE
+ */
+fih_int boot_secure_memequal(const void *s1, const void *s2, size_t n)
```
I can see what you're doing but when I see `secure_memequal` I expect a function that is constant time, which is not the case here, maybe a more appropriate name would be helpful, maybe settle on `fih` like `boot_fih_memequal` (sorry for the lack of inspiration!)

utzig commented on boot/bootutil/src/bootutil_misc.c:
```diff
@@ -107,6 +108,38 @@ static const struct boot_swap_table boot_swap_tables[] = {
+
+    for (i = 0; i < n; i++) {
+        if (s1_p[i] != s2_p[i]) {
+            goto out;;
```
can remove one `;`

utzig commented:
This should get back around 40 bytes for someone who doesn't want to have the fault-injection protections:

```
diff --git a/boot/bootutil/src/bootutil_misc.c b/boot/bootutil/src/bootutil_misc.c
index 9d80b5a06..a48b1431f 100644
--- a/boot/bootutil/src/bootutil_misc.c
+++ b/boot/bootutil/src/bootutil_misc.c
@@ -120,6 +120,12 @@ static const struct boot_swap_table boot_swap_tables[] = {
  *
  * @return      FIH_SUCCESS if memory regions are equal, otherwise FIH_FAILURE
  */
+#ifdef MCUBOOT_FIH_PROFILE_OFF
+inline int boot_secure_memequal(const void *s1, const void *s2, size_t n)
+{
+    return memcmp(s1, s2, n);
+}
+#else
 fih_int boot_secure_memequal(const void *s1, const void *s2, size_t n)
 {
     size_t i;
@@ -139,6 +145,7 @@ fih_int boot_secure_memequal(const void *s1, const void *s2, size_t n)
 out:
     FIH_RET(ret);
 }
+#endif
 
 static int
 boot_magic_decode(const uint32_t *magic)
```

nvlsianpu commented on boot/bootutil/src/image_validate.c:
```diff
@@ -449,44 +464,55 @@ bootutil_img_validate(struct enc_key_data *enc_state, int image_index,
-            if (img_security_cnt < security_cnt) {
-                /* The image's security counter is not accepted. */
-                return -1;
+            fih_rc = fih_int_encode_zero_equality(img_security_cnt <
```
```suggestion
  if (img_security_cnt < fih_int_decode(security_cnt)) {
    fih_rc = FIH_SUCCESS;
  } else {
    fih_rc = FIH_FAILURE;
  }
```
just think that usage _encode_ funct here is overcomplicated

tamasban replied:
Yeah, tend to agree.  I think both version has the same level of resistance against FI. With the usage of `fih_int_encode_zero_equality` the code is a bit more concise and if  `fih_int_encode_zero_equality` would have adding some random delay (calling `fih_delay()`) then it would be more robust. But I can accept your version as well. Let me know which one do you prefer.

nvlsianpu replied:
This was not a strong suggestion from my side.

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -386,13 +387,14 @@ boot_write_status(const struct boot_loader_state *state, struct boot_status *bs)
     TARGET_STATIC uint8_t tmpbuf[BOOT_TMPBUF_SZ];
     uint8_t image_index;
-    int rc;
+    int rc = -1;
```
```suggestion
    int rc;
```

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -386,13 +387,14 @@ boot_write_status(const struct boot_loader_state *state, struct boot_status *bs)
     uint8_t image_index;
-    int rc;
+    int rc = -1;
+    fih_int fih_rc = FIH_FAILURE;
```
```suggestion
    fih_int fih_rc ;
```
always assigned by FIH_CALL(bootutil_img_validate...

tamasban replied:
It is intended to protect against attack when the PC is tried to abuse such a way (manipulating lower bits with EM pulse) to jump at the end of the function and just returns without performing any operation. Very unlikely and hard to achieve in practice. But during a certification process, code reviewer can mandate to have negative return value set by default.

utzig replied:
Yes, all variables should be initialized to the error value, otherwise abuse and unintended early exit would result in success returns. I wonder if those could not just be initialized at the beginning of the function instead of where they were declared, but I don't have strong opinions about it anymore.

nvlsianpu replied:
Ok, I'm fine with this clarification - although all that might be compiled out as optimization.

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -574,19 +575,20 @@ boot_version_cmp(const struct image_version *ver1,
     int area_id;
-    int rc;
+    fih_int fih_rc = FIH_FAILURE;
+    int rc = 0;
```
```suggestion
    int rc ;
```

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -643,16 +646,14 @@ boot_validate_slot(struct boot_loader_state *state, int slot,
     }
 
-    /* Image in the secondary slot is valid. */
-    rc = 0;
```
shouldn't a success code be encoded to `fih_rc`?

tamasban replied:
You are right, and `fih_int_encode()` does exactly this: Converting an int to fih_int. You can check: fault_injection_hardening.h:203

nvlsianpu replied:
I see` line 637:  FIH_CALL(boot_image_check, fih_rc, state, hdr, fap, bs);` may set `fih_rc` to FIH_SUCCESS.

So this function now returns either `FIH_FAILURE` (previous `-1`), `FIH_SUCCESS` (previous `0`), and the value which is encoded by fih_int_encode(1) call (previous `1`).

I can't see initialization for either `FIH_FAILURE` and `FIH_SUCCESS`, so what the warranty that all value listed here are distinguishable?

tamasban replied:
FIH_FAILURE and FIH_SUCCESS are just normal integer values if MCUBOOT_FIH_PROFILE is OFF. Actually their value is 0 and 1 in this case. Complex value are used only if some level of FI protection is applied.  In this case might the returns values are colliding: FIH_FAILURE  and 1 (no bootable image). will double check.

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -675,7 +676,7 @@ boot_update_security_counter(uint8_t image_index, int slot,
     const struct flash_area *fap = NULL;
     uint32_t img_security_cnt;
-    int rc;
+    int rc = 0;
```
```suggestion
    int rc;
```

nvlsianpu commented on boot/bootutil/src/loader.c:
```diff
@@ -574,19 +575,20 @@ boot_version_cmp(const struct image_version *ver1,
  *         1 if no bootloable image was found
```
@returns description need to be upadted.

nvlsianpu commented on boot/bootutil/include/bootutil/fault_injection_hardening.h:
```diff
@@ -0,0 +1,356 @@
+ * common value for a corrupted register so complex bit-patterns are used
+ */
+#ifndef MCUBOOT_FIH_PROFILE_OFF
+#define FIH_POSITIVE_VALUE 0x1AAAAAAA
```
so these two value are problematic if would be a case as previous integer return-value, not a case for this code, but hypothetically might be an issue.

nvlsianpu replied:
I men that `FIH_POSITIVE_VALUE== fih_int_encoded(0x1AAAAAAA)`...

tamasban replied:
Not really got it. Please elaborate.

nvlsianpu replied:
`FIH_POSITIVE_VALUE supersedes what before was `0`. FIH_NEGATIV_VALUE` supersedes what was an negative error-code before.
If an interface will try to use return value in a manner where encodes a range of integers as successful results, and uses `FIH_NEGATIV_VALUE` for encoding an error - this will cause misunderstanding, and probably an issue.

fih_int with either `FIH_NEGATIV_VALUE` and `FIH_POSITIVE_VALUE`  should be use like boolean type.

tamasban replied:
Arbitrary error code can be returned, you can use `fih_int_encode()` and `fih_int_decode` to covert them.
There is an example in: loader.c:726
If MCUBOOT_FIH_PROFILE is OFF then these are just normal integers. I have updated the PR to preserve the original error codes. So distinction can be done on the application layer (where `boot_go` or `split_go` returns) .

utzig commented on boot/zephyr/Kconfig:
```diff
@@ -288,6 +288,45 @@ config BOOT_SHARE_DATA
+	  delays to make code execution less predictable. Random delays requires an
+	  entropy source.
+
+endcoice
```
typo: `endchoice`

nvlsianpu commented on boot/zephyr/Kconfig:
```diff
@@ -297,6 +297,45 @@ config BOOT_SHARE_DATA
+
+config BOOT_FIH_PROFILE_OFF
+	bool "No hardening against hardware level fault injection"
+	select BOOT_FIH_PROFILE_OFF
```
```suggestion
```

nvlsianpu commented on boot/zephyr/Kconfig:
```diff
@@ -297,6 +297,45 @@ config BOOT_SHARE_DATA
+
+config BOOT_FIH_PROFILE_LOW
+	bool "Moderate level hardening against hardware level fault injection"
+	select BOOT_FIH_PROFILE_LOW
```
```suggestion
```

nvlsianpu commented on boot/zephyr/Kconfig:
```diff
@@ -297,6 +297,45 @@ config BOOT_SHARE_DATA
+
+config BOOT_FIH_PROFILE_MEDIUM
+	bool "Medium level hardening against hardware level fault injection"
+	select BOOT_FIH_PROFILE_MEDIUM
```
```suggestion
```

nvlsianpu commented on boot/zephyr/Kconfig:
```diff
@@ -297,6 +297,45 @@ config BOOT_SHARE_DATA
+
+config BOOT_FIH_PROFILE_HIGH
+	bool "Maximum level hardening against hardware level fault injection"
+	select BOOT_FIH_PROFILE_HIGH
```
```suggestion
```

nvlsianpu reviewed (commented):
remove selecting kconfig property by itself

utzig commented:
> remove selecting kconfig property by itself

Yes, and there are other issues, I will squash in fixes.

tamasban commented:
> > remove selecting kconfig property by itself
> 
> Yes, and there are other issues, I will squash in fixes.

Thanks, for the fixes @utzig!

utzig reviewed (approved):
It looks OK, I am a bit bothered that it increases all binaries by 100 bytes or so, which is not that bad but would be better to not add anything if set to `_OFF`. Also for Tinycrypt users `uECC_vli_equal` has to be updated and same for fiat's `CRYPTO_memcmp`, but it can be done in a follow up PR before releasing the next version.

utzig commented:
@romanjoe Are you able to check this PR for Cypress today?

romanjoe commented:
@utzig yes, i'll do right now

romanjoe commented:
First issue i have faced while trying to compile is error message:
`mcuboot/boot/cypress/../bootutil/include/bootutil/crypto/ecdsa_p256.h:17:6: error: #error "One crypto backend must be defined: either CC310 or TINYCRYPT"
     #error "One crypto backend must be defined: either CC310 or TINYCRYPT"`

cypress use `#define MCUBOOT_USE_MBED_TLS`

I assume this change is introduced with https://github.com/JuulLabs-OSS/mcuboot/pull/798, but we did not rebase our internal developing to include this PR yet, so i assume this issue is present in master as well, i need to resolve this issue first

utzig commented:
@romanjoe Please drop your commits and afterwards @tamasban can rebase.

romanjoe commented:
This PR fixes compilation for cypress on master: https://github.com/JuulLabs-OSS/mcuboot/pull/825 - please rebase
@utzig @tamasban i also added additional fix regarding cypress in this PR

tamasban commented:
> This PR fixes compilation for cypress on master: #825 - please rebase
> @utzig @tamasban i also added additional fix regarding cypress in this PR

I have squashed your fix and did the rebase.

romanjoe commented:
@utzig @tamasban i checked latest with cypress and it works fine, thank you
